crash on browsing lots of movies and backing out

using the same technique to try and reproduce the crash in #202 with the PR #203 code, i got this crash:
```
FATAL EXCEPTION: main
Process: org.amahi.anywhere, PID: 28899
java.util.concurrent.RejectedExecutionException: Task android.os.AsyncTask$3@7d44d9a rejected from java.util.concurrent.ThreadPoolExecutor@42626cb[Running, pool size = 9, active threads = 9, queued tasks = 128, completed tasks = 735]
    at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2014)
    at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:794)
    at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1340)
    at android.os.AsyncTask.executeOnExecutor(AsyncTask.java:607)
    at org.amahi.anywhere.task.FileMetadataRetrievingTask.execute(FileMetadataRetrievingTask.java:56)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.bindFileMetadataView(ServerFilesMetadataAdapter.java:127)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.bindView(ServerFilesMetadataAdapter.java:88)
    at org.amahi.anywhere.adapter.FilesFilterBaseAdapter.getView(FilesFilterBaseAdapter.java:103)
    at android.widget.AbsListView.obtainView(AbsListView.java:2421)
    at android.widget.GridView.makeAndAddView(GridView.java:1440)
    at android.widget.GridView.makeRow(GridView.java:366)
    at android.widget.GridView.fillDown(GridView.java:307)
    at android.widget.GridView.fillGap(GridView.java:267)
    at android.widget.AbsListView.trackMotionScroll(AbsListView.java:5283)
    at android.widget.AbsListView$FlingRunnable.run(AbsListView.java:4831)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:873)
    at android.view.Choreographer.doCallbacks(Choreographer.java:676)
    at android.view.Choreographer.doFrame(Choreographer.java:603)
    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:859)
    at android.os.Handler.handleCallback(Handler.java:739)
    at android.os.Handler.dispatchMessage(Handler.java:95)
    at android.os.Looper.loop(Looper.java:168)

```

This area is rich in crashes ... same thing ... browse, back, back ... tap, crash ..

```
FATAL EXCEPTION: main
Process: org.amahi.anywhere, PID: 14738
java.lang.RuntimeException: Could not dispatch event: class org.amahi.anywhere.bus.FileMetadataRetrievedEvent to handler [EventHandler public void org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.onFileMetadataRetrieved(org.amahi.anywhere.bus.FileMetadataRetrievedEvent)]: You cannot start a load for a destroyed activity
    at com.squareup.otto.Bus.throwRuntimeException(Bus.java:460)
    at com.squareup.otto.Bus.dispatch(Bus.java:387)
    at com.squareup.otto.Bus.dispatchQueuedEvents(Bus.java:368)
    at com.squareup.otto.Bus.post(Bus.java:337)
    at org.amahi.anywhere.task.FileMetadataRetrievingTask.onResponse(FileMetadataRetrievingTask.java:74)
    at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$1.run(ExecutorCallAdapterFactory.java:68)
    at android.os.Handler.handleCallback(Handler.java:739)
    at android.os.Handler.dispatchMessage(Handler.java:95)
    at android.os.Looper.loop(Looper.java:168)
    at android.app.ActivityThread.main(ActivityThread.java:5885)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:797)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:687)
 Caused by: java.lang.IllegalArgumentException: You cannot start a load for a destroyed activity
    at com.bumptech.glide.manager.RequestManagerRetriever.assertNotDestroyed(RequestManagerRetriever.java:134)
    at com.bumptech.glide.manager.RequestManagerRetriever.get(RequestManagerRetriever.java:102)
    at com.bumptech.glide.manager.RequestManagerRetriever.get(RequestManagerRetriever.java:87)
    at com.bumptech.glide.Glide.with(Glide.java:629)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.bindFileMetadataView(ServerFilesMetadataAdapter.java:156)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.bindView(ServerFilesMetadataAdapter.java:145)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.onFileMetadataRetrieved(ServerFilesMetadataAdapter.java:137)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.squareup.otto.EventHandler.handleEvent(EventHandler.java:89)
    at com.squareup.otto.Bus.dispatch(Bus.java:385)
    at com.squareup.otto.Bus.dispatchQueuedEvents(Bus.java:368) 
    at com.squareup.otto.Bus.post(Bus.java:337) 
    at org.amahi.anywhere.task.FileMetadataRetrievingTask.onResponse(FileMetadataRetrievingTask.java:74) 
    at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$1.run(ExecutorCallAdapterFactory.java:68) 
    at android.os.Handler.handleCallback(Handler.java:739) 
    at android.os.Handler.dispatchMessage(Handler.java:95) 
    at android.os.Looper.loop(Looper.java:168) 
    at android.app.ActivityThread.main(ActivityThread.java:5885) 
    at java.lang.reflect.Method.invoke(Native Method) 
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:797) 
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:687) 
```

--- a/src/main/java/org/amahi/anywhere/adapter/ServerFilesMetadataAdapter.java
+++ b/src/main/java/org/amahi/anywhere/adapter/ServerFilesMetadataAdapter.java
@@ -124,7 +124,7 @@ public class ServerFilesMetadataAdapter extends FilesFilterBaseAdapter
 		fileView.setTag(Tags.SHARE, serverShare);
 		fileView.setTag(Tags.FILE, file);
 
-		FileMetadataRetrievingTask.execute(serverClient, fileView);
+		new FileMetadataRetrievingTask(serverClient, fileView).execute();
 	}
 
 	@Subscribe

--- a/src/main/java/org/amahi/anywhere/adapter/ServerFilesMetadataAdapter.java
+++ b/src/main/java/org/amahi/anywhere/adapter/ServerFilesMetadataAdapter.java
@@ -85,7 +85,11 @@ public class ServerFilesMetadataAdapter extends FilesFilterBaseAdapter
 		if (Mimes.match(file.getMime()) != Mimes.Type.VIDEO) {
 			bindFileView(file, fileView);
 		} else {
-			bindFileMetadataView(file, fileView);
+			if(!file.isMetaDataFetched()) {
+				bindFileMetadataView(file, fileView);
+			} else {
+				bindView(file, file.getFileMetadata(), fileView);
+			}
 		}
 		if (Mimes.match(file.getMime()) == Mimes.Type.IMAGE) {
 			setUpImageIcon(file, fileIcon);
@@ -103,7 +107,7 @@ public class ServerFilesMetadataAdapter extends FilesFilterBaseAdapter
 		fileIcon.setBackgroundResource(R.color.background_secondary);
 	}
 
-	private static void bindFileView(ServerFile file, View fileView) {
+	private void bindFileView(ServerFile file, View fileView) {
 		TextView fileTitle = (TextView) fileView.getTag(Tags.FILE_TITLE);
 		ImageView fileIcon = (ImageView) fileView.getTag(Tags.FILE_ICON);
 
@@ -129,6 +133,7 @@ public class ServerFilesMetadataAdapter extends FilesFilterBaseAdapter
 
 	@Subscribe
 	public void onFileMetadataRetrieved(FileMetadataRetrievedEvent event) {
+		event.getFile().setMetaDataFetched(true);
 		bindView(event.getFile(), event.getFileMetadata(), event.getFileView());
 	}
 
@@ -136,6 +141,7 @@ public class ServerFilesMetadataAdapter extends FilesFilterBaseAdapter
 		if (fileMetadata == null) {
 			bindFileView(file, fileView);
 		} else {
+			file.setFileMetadata(fileMetadata);
 			bindFileMetadataView(file, fileMetadata, fileView);
 		}
 	}

--- a/src/main/java/org/amahi/anywhere/server/client/ServerClient.java
+++ b/src/main/java/org/amahi/anywhere/server/client/ServerClient.java
@@ -51,6 +51,7 @@ import java.io.IOException;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 
+import retrofit2.Callback;
 
 
 /**
@@ -225,17 +226,11 @@ public class ServerClient
 			.build();
 	}
 
-	public ServerFileMetadata getFileMetadata(ServerShare share, ServerFile file) {
+	public void getFileMetadata(ServerShare share, ServerFile file, Callback<ServerFileMetadata> callback) {
         if ((server == null) || (share == null) || (file == null)){
-            return null;
+            return;
         }
-
-		try {
-			return serverApi.getFileMetadata(server.getSession(), file.getName(), share.getTag()).execute().body();
-		} catch (IOException e) {
-			e.printStackTrace();
-			return null;
-		}
+		serverApi.getFileMetadata(server.getSession(), file.getName(), share.getTag()).enqueue(callback);
 	}
 
 	public void getApps() {

--- a/src/main/java/org/amahi/anywhere/server/model/ServerFile.java
+++ b/src/main/java/org/amahi/anywhere/server/model/ServerFile.java
@@ -43,14 +43,18 @@ public class ServerFile implements Parcelable
 	@SerializedName("mtime")
 	private Date modificationTime;
 
-       @SerializedName("size")
-       private long size;
+    @SerializedName("size")
+    private long size;
 
-       public long getSize() {
+    private ServerFileMetadata fileMetadata;
+
+	private boolean isMetaDataFetched;
+
+    public long getSize() {
             return size;
        }
 
-       public void setParentFile(ServerFile parentFile) {
+    public void setParentFile(ServerFile parentFile) {
 		this.parentFile = parentFile;
 	}
 
@@ -68,6 +72,22 @@ public class ServerFile implements Parcelable
 
 	public Date getModificationTime() {
 		return modificationTime;
+	}
+
+    public ServerFileMetadata getFileMetadata() {
+        return fileMetadata;
+    }
+
+    public void setFileMetadata(ServerFileMetadata fileMetadata) {
+        this.fileMetadata = fileMetadata;
+    }
+
+	public boolean isMetaDataFetched() {
+		return isMetaDataFetched;
+	}
+
+	public void setMetaDataFetched(boolean metaDataFetched) {
+		isMetaDataFetched = metaDataFetched;
 	}
 
 	public String getPath() {

--- a/src/main/java/org/amahi/anywhere/task/FileMetadataRetrievingTask.java
+++ b/src/main/java/org/amahi/anywhere/task/FileMetadataRetrievingTask.java
@@ -19,7 +19,6 @@
 
 package org.amahi.anywhere.task;
 
-import android.os.AsyncTask;
 import android.view.View;
 
 import org.amahi.anywhere.adapter.ServerFilesMetadataAdapter;
@@ -34,36 +33,32 @@ import org.amahi.anywhere.server.model.ServerShare;
 import java.lang.ref.Reference;
 import java.lang.ref.WeakReference;
 
-public class FileMetadataRetrievingTask extends AsyncTask<Void, Void, ServerFileMetadata> {
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
 
-    private final ServerClient serverClient;
+public class FileMetadataRetrievingTask implements Callback<ServerFileMetadata> {
 
     private final Reference<View> fileViewReference;
 
     private final ServerShare share;
     private final ServerFile file;
+    private final ServerClient serverClient;
 
-    private FileMetadataRetrievingTask(ServerClient serverClient, View fileView) {
+    public FileMetadataRetrievingTask(ServerClient serverClient, View fileView) {
         this.serverClient = serverClient;
-
         this.fileViewReference = new WeakReference<>(fileView);
 
         this.share = (ServerShare) fileView.getTag(ServerFilesMetadataAdapter.Tags.SHARE);
         this.file = (ServerFile) fileView.getTag(ServerFilesMetadataAdapter.Tags.FILE);
     }
 
-    public static void execute(ServerClient serverClient, View fileView) {
-        new FileMetadataRetrievingTask(serverClient, fileView).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+    public void execute() {
+        serverClient.getFileMetadata(share, file, this);
     }
 
     @Override
-    protected ServerFileMetadata doInBackground(Void... parameters) {
-        return serverClient.getFileMetadata(share, file);
-    }
-
-    @Override
-    protected void onPostExecute(ServerFileMetadata fileMetadata) {
-        super.onPostExecute(fileMetadata);
+    public void onResponse(Call<ServerFileMetadata> call, Response<ServerFileMetadata> response) {
 
         View fileView = fileViewReference.get();
 
@@ -75,8 +70,12 @@ public class FileMetadataRetrievingTask extends AsyncTask<Void, Void, ServerFile
             return;
         }
 
-        BusEvent busEvent = new FileMetadataRetrievedEvent(file, fileMetadata, fileView);
+        BusEvent busEvent = new FileMetadataRetrievedEvent(file, response.body(), fileView);
         BusProvider.getBus().post(busEvent);
+    }
+
+    @Override
+    public void onFailure(Call<ServerFileMetadata> call, Throwable t) {
 
     }
 }

